### DISCUSSION TITLE
tutorials/voyage-part1/12-referencing-an-entity

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#79] [course feedback]
# tutorials/voyage-part1/12-referencing-an-entity

Reference an entity from another subgraph, and learn how the router resolves entities

https://www.apollographql.com/tutorials/voyage-part1/12-referencing-an-entity

**Comment by Aimeana100** [comment]
Hello, I have an issue: I am still getting null locations after setting resorlvers.Review resolver.
Can anyone assist me, please?

**Reply by lizhennessy** [reply]
Hi there @aimeana100! I am happy to help. Do you have your code pushed up to a repository that I could take a look at?

**Reply by tsinghammer** [reply]
@lizhennessy I'm having the same issue and cannot find what's missing. What did it turn out to be @Aimeana100 's case?
Otherwise I can also push my repo somewhere.

**Reply by tsinghammer** [reply]
PS: in my case the Location's __resolveReference function is never hit, the Review's location resolver function however is and logs the correct id.

**Reply by tsinghammer** [reply]
Ok, in my case I was adding the reference resolver in the wrong file, the one in the final folder, which I did not recognize.
Sorry for spamming this thread.

**Reply by rodmirsantana** [reply]
I also still am getting `null` locations on the query response after updating the reviews subgraph schema changes. What could have gone wrong?

**Reply by rodmirsantana** [reply]
@lizhennessy @mabuyo

**Reply by mabuyo** [reply]
@rodmirsantana I would check if your resolvers are defined correctly and that they are getting called (console.logs can help here, or you can debug breakpoints). Feel free to send a link to your repo and we can try and troubleshoot that way as well.

**Comment by AnatholeTech** [comment]
Hello @lizhennessy, I repeated all the steps, and the issue got resolved; it was probably one of the instructions I wrongly applied.

**Comment by Dark417** [comment]
As I posted in 'Connecting data using entities', I manually downloaded the supergraph and ran .router --dev --supergraph xxx.graphql
Now again, I published review subgraph and can see the change in studio, but port 4000 can't pick up the change. I did the manual process again and finally saw name resolved under location.
I went back and forth and can't find the step I missed, how can I set it up so supergraph and router updates automatically?

**Reply by Dark417** [reply]
Update:
Use $env:APOLLO_KEY = ".." it works

**Comment by frankielin2015** [comment]
Hello there,
I have a question regarding passing input data from subgraph A to subgraph B
let's say I have a query

query sampleQuery {
 review(reviewId: 2) {
   reviewId,
   reviewDate
   location {
   name
   }
  }
}

Let's say for some reason I need access that reviewId from Review subgraph in Location ____resolveReference() how can we do that?

**Reply by mabuyo** [reply]
Hey there @frankielin2015 - apologies for the delayed response. So to clarify, `Location.__resolveReference`, or the reference resolver, only has access to the Location entity representation. An entity representation is object that contains the entity's @key fields, plus its __typename field. So you wouldn't be able to access the `reviewId`. If you need that property to resolve the entity, then, it should be part of the Location type's primary key fields.

**Comment by TrillionXwin** [comment]
The name field from location does not resolve, id does. the two additional fields from later on also resolve.
I checked against the final version in github, my code is the same. I even copy pasted the exact code from your repository into the graphql files and resolvers. I restarted everything and updated the resolvers, still does not work.
The final webpage works.

Are you sure this is a valid query? Wouldn't we need to add location's name field to the reviews resolver to get it to contribute to the reviews subgraph?
```
query GetLatestReviews {
  latestReviews {
    id
    comment
    rating
    location {
      name
    }
  }
}
```

Creating a resolveReference on key field does not automatically give access to all of the fields in the type. It is not like an inner join.
How can I use location id to return the name field from location? Granted, it's not necessary.
I guess in this case we can make the name a secondary key field and

**Reply by lizhennessy** [reply]
Hi there @TrillionXwin! In this case, we can break down the parts of the query between the subgraphs that are responsible for fulfilling them.

From `reviews`, we want to get `latestReviews`; for each latest review, we're requesting its `id`, `comment`, and `rating`, and all of this is fine and easy for the `reviews` subgraph to provide. When we get to `location`, however, all we have in our `reviews` subgraph is a `locationId`. This is why we set up a `__resolveReference` function, so that we can use the `locationId` from the `reviews` subgraph and correlate it with the correct data from the `locations` subgraph.

```
query GetLatestReviews {
  latestReviews {  
    id    # provided by reviews
    comment    # provided by reviews
    rating    # provided by reviews
    location {
      name    # provided by locations
    }
  }
}
```

When we pass along a `locationId` from `reviews`, such as `loc-1`, this ID can be conveyed to the `locations` subgraph's `__resolveReference` function. Using this ID, it looks up the correct `Location` object in its database, and can then provide all of the other fields, including `name`! 

We use the `__resolveReference` function to "link" up data between the `reviews` and `locations` subgraphs regarding a specific `Location`. The goal here is to find the key that they have in common (a location's ID) so that it can be used to pluck the correct data from both locations! So to make this query work, all we need is the ID field they have in common.

I hope this explanation helps! Please let us know if you have any other questions.

**Comment by twm-faraujo** [comment]
"""
2024-05-14T21:29:33.727898Z INFO fetch{apollo.subgraph.name=locations,} execution{graphql.operation.type=query,} supergraph{graphql.operation.name=LatestReviews,graphql.document=query LatestReviews {
  latestReviews {
    rating
    comment
    id
    location {
      name
    }
  }
},graphql.operation.name=LatestReviews,} router{http.method=POST,http.request.method=POST,http.route=/,http.flavor=HTTP/1.1,trace_id=a2ee760f1d1438b27d11cedd0e437fe8,url.path=/,client.name=,client.version=,} request{http.method=POST,http.request.method=POST,http.route=/,http.flavor=HTTP/1.1,}  redacted 

"""
Does anybody knows why is this error?

**Reply by twm-faraujo** [reply]
Well, I think I figure it out, I'm working on WSL and I have to run "sudo ufw allow
 4001" I did it for 4000 and 4002 too
